src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs

```diff
@@ -39,20 +34,50 @@ internal sealed class RegexFCD
         internal const int Boundary = 0x0040;
         internal const int ECMABoundary = 0x0080;
 
-        /*
-         * This is the one of the only two functions that should be called from outside.
-         * It takes a RegexTree and computes the set of chars that can start it.
-         */
-        internal static RegexPrefix FirstChars(RegexTree t)
+        private readonly ValueListBuilder<RegexFC> _fcStack;
+        private readonly ValueListBuilder<int> _intStack;
+        private bool _skipAllChildren;      // don't process any more children at the current level
+        private bool _skipchild;            // don't process the current child.
+        private bool _failed;
+
+        private RegexFCD(ValueListBuilder<RegexFC> fcStack, ValueListBuilder<int> intStack)
         {
-            RegexFCD s = new RegexFCD();
-            RegexFC fc = s.RegexFCFromRegexTree(t);
+            _fcStack = fcStack;
+            _intStack = intStack;
+            _failed = false;
+            _skipchild = false;
+            _skipAllChildren = false;
+        }
 
-            if (fc == null || fc._nullable)
+        /// <summary>
+        /// This is the one of the only two functions that should be called from outside.
+        /// It takes a RegexTree and computes the set of chars that can start it.
+        /// </summary>
+        internal static RegexPrefix? FirstChars(RegexTree t)
+        {
+            // Create/rent buffers
+            Span<int> intSpan = stackalloc int[StackBufferSize];
+            var intStack = new ValueListBuilder<int>(intSpan);
+            RegexFC[] fcSpan = ArrayPool<RegexFC>.Shared.Rent(StackBufferSize);
+            var fcStack = new ValueListBuilder<RegexFC>(fcSpan);
+
+            RegexFCD s = new RegexFCD(fcStack, intStack);
+            RegexFC? fc = s.RegexFCFromRegexTree(t);
+
+            // Return rented buffers. Clear RegexFC buffer manually as only
+            // the reference to the RegexCharClass needs to be cleared.
+            for (int i = 0; i < fcStack.Length; i++)
+                fcStack[i].Dispose();
```

Comment:
Each of these fcStacks has a rented buffer?  How many of them are there?  If we're consuming lots of ArrayPool buffers for one regex operation, what happens when regexs are performed in parallel, such as in an ASP.NET application with lots of requests all being processed concurrently?

Reply:
Nevermind, I misunderstood, and I see further on that they don't wrap rented buffers.

Reply:
still a good question! That code is only called once per Regex construction, it's part of the pre-processing stuff (tree, opcode generation, ...). Of course, if multiple instances of Regex are generated while one of the arrays is rented and not yet returned the ArrayPool will allocate a new array with size 32. Not get rid off the unused memory after time @jkotas thinks about integrating the ArrayPool with the GC, described here: https://github.com/dotnet/coreclr/issues/16502#issuecomment-367720105

Reply:
Have you measured that you are really getting any speedup from using pooled arrays for these? The GC is very good at getting rid of small short-term garbage that this is producing. I think that we are going overboard with trying to pool here.

Reply:
Throughput stays same, allocation reduction is the only benefit by using ArrayPools for RegexFCD. With commit 1 I was using a list with an initial size of 32 which results in same throughput but 40MB more allocation. I don't have a strong opinion here and I see that change more as an experiment. I was talking with Stephen about it offline beforehand and we concluded that it is worth giving it a try. As said, I'm absolutely fine with using a List here for the RegexFCD data store.

Reply:
Then I would not be pooling here. I think allocations during regex construction are ok, it is not something that we should go out of our way to optimize.